torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3721,9 +3702,7 @@ def sharded_optim_state_dict(
     def shard_full_optim_state_dict(
         full_optim_state_dict: Dict[str, Any],
         model: torch.nn.Module,
-        optim_input: Optional[Union[
-            List[Dict[str, Any]], Iterable[torch.nn.Parameter],
-        ]] = None,
+        optim: torch.optim.Optimizer,
```

Comment:
curious why we're getting rid of `None` default codepath and enforcing the arg?

Reply:
Similarly, do we need `full_optim_state_dict` when it's just `optim.state_dict()`??

Reply:
I do not believe that `full_optim_state_dict()` is the same as `optim.state_dict()`.

Each `optim` is a per-rank local optimizer that operates in terms of sharded `FlatParameter`s. The return value of `full_optim_state_dict()` is in terms of unsharded original parameters.